### PR TITLE
Introduce ChoicesContext

### PR DESCRIPTION
**Pull request by djhi**
- [x] Refactor Reference inputs and their controllers to add the ChoicesContext
- [x] Ensure SelectInput can use it
- [x] Ensure SelectArrayInput can use it
- [x] Ensure AutocompleteInput can use it
- [x] Ensure AutocompleteArrayInput can use it
- [x] Update examples
- [x] Add ability to filter the ChoicesContext data (all choices, selected choices or available choices)
- [x] Upgrade guide
- [x] Documentation

### CODE REVIEW

File: examples/simple/src/posts/TagReferenceInput.tsx [hunk 1]
```diff
@@ -58,33 +46,27 @@ const TagReferenceInput = ({
...
-                {...props}
-                filter={filter}
-                filterToQuery={filterToQuery}
-            >
-                <AutocompleteArrayInput
-                    create={<CreateTag />}
-                    optionText={`name.${locale}`}
-                />
+            <ReferenceArrayInput {...props} perPage={5} filters={published}>
+                <DatagridInput>
```
Comment (djhi) [inline examples/simple/src/posts/TagReferenceInput.tsx:60]: I'll revert this when I'm done

**Review by WiXSL — CHANGES_REQUESTED** [review]
`ReferenceInput.tsx` still has `filterToQuery` prop referenced in the code

File: UPGRADE.md [hunk 2]
```diff
@@ -2798,3 +2798,96 @@ import { BooleanInput } from 'react-admin';
...
+### Other Changes
+
+The other changes only impact you if you had a custom child component for either the `<ReferenceInput>` or `<ReferenceArrayInput>`.
+
+The `ChoicesContext` provides access to the choices via 3 properties:
+- `availableChoices`: The choices which are not selected but matches the parameters (sorting, pagination and filters)
+- `selectedChoices`: The selected choices. 
+- `allChoices`: Merge of both available and selected choices. 
+
+The `ReferenceInput` and `ReferenceArrayInput` components also handled the form binding and injected final-form `input` and `meta` to their child. They don't need to do that anymore and simply watch the form value changes ti ensure they load the required data when needed.
```
Comment (WiXSL) [inline UPGRADE.md:2873]:
```suggestion
The `ReferenceInput` and `ReferenceArrayInput` components also handled the form binding and injected final-form `input` and `meta` to their child. They don't need to do that anymore, and simply watch the form value changes to ensure they load the required data when needed.
```

**Review by WiXSL — CHANGES_REQUESTED** [review]

File: UPGRADE.md [hunk 3]
```diff
@@ -2798,3 +2798,96 @@ import { BooleanInput } from 'react-admin';
...
+    >
+-        <AutocompleteInput />
++        <AutocompleteInput filterToQuery={search => ({ email: search })} />
+    </ReferenceInput>
+]
+```
+
+### `<ReferenceArrayInput>` No Longer Provides a `ListContext`
+
+As the `ChoicesContext` now provide an API very similar to the `ListContext`, it no longer sets up a `ListContext`. If you used this to display a `<Datagrid>`, we will soon provide the `<DatagridInput>` for this purpose.
```
Comment (WiXSL) [inline UPGRADE.md:2824]: 😄 I like this futuristic comment

Comment (djhi) [reply UPGRADE.md:2824]: There's a WIP version in this PR already!

**Review by WiXSL — CHANGES_REQUESTED** [review]

**Review by WiXSL — COMMENTED** [review]
Partial review, I'll continue tomorrow

File: docs/ReferenceArrayInput.md [hunk 4]
```diff
@@ -25,9 +25,9 @@ http://myapi.com/tags?id=[1,23,4]
 http://myapi.com/tags?page=1&perPage=25
 ```
 
-Once it receives the deduplicated reference resources, this component delegates rendering to a subcomponent, by providing the possible choices through the `ReferenceArrayInputContext`. This context value can be accessed with the [`useReferenceArrayInputContext`](#usereferencearrayinputcontext) hook.
+Once it receives the deduplicated reference resources, this component delegates rendering its child component, by providing the possible choices through the `ChoicesContext`. This context value can be accessed with the [`useChoicesContext`](./useChoicesContext.md) hook.
```
Comment (fzaninotto) [inline docs/ReferenceArrayInput.md:28]:
```suggestion
Once it receives the deduplicated reference resources, this component delegates rendering to its child component, by providing the possible choices through the `ChoicesContext`. This context value can be accessed with the [`useChoicesContext`](./useChoicesContext.md) hook.
```

File: docs/useChoicesContext.md [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+import { AutocompleteInput, useChoicesContext } from 'react-admin';
+
+export const PostInput = (props) => {
+    const { setFilters } = useChoicesContext();
+
+    const handleCheckboxChange = (event, checked) => {
+        setFilters({ published: checked });
+    };
+
+    return (
```
Comment (fzaninotto) [inline docs/useChoicesContext.md:31]: 👍  Excellent example!

File: packages/ra-core/src/form/choices/ChoicesContext.ts [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+
+/**
+ * Context to store choices and functions to retrieve them.
+ *
+ * Use the useChoicesContext() hook to read the context.
+ */
+export const ChoicesContext = createContext<ChoicesContextValue>(undefined);
+
+export type ChoicesContextValue<RecordType extends RaRecord = any> = {
+    sort: SortPayload;
```
Comment (fzaninotto) [inline packages/ra-core/src/form/choices/ChoicesContext.ts:12]: could you please sort the keys alphabetically, or add comments to explain the sort logic?

File: packages/ra-core/src/form/choices/useChoicesContext.ts [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+            setFilters: options.setFilters ?? list.setFilters,
+            setPage: options.setPage ?? list.setPage,
+            setPerPage: options.setPerPage ?? list.setPerPage,
+            setSort: options.setSort ?? list.setSort,
+            showFilter: options.showFilter ?? list.showFilter,
+            sort: options.sort ?? list.sort,
+            source: options.source,
+            total: options.total ?? list.total,
+        }),
+        [data, list, options]
```
Comment (fzaninotto) [inline packages/ra-core/src/form/choices/useChoicesContext.ts:41]: you're not using the same logic as for other context hooks. If a prop is passed as argument, it should override the context value, even if there is a context

File: packages/ra-core/src/form/choices/useChoicesContext.ts [hunk 8]
```diff
@@ -0,0 +1,48 @@
+import { useContext, useMemo } from 'react';
+import { RaRecord } from '../../types';
+import { useList } from '../../controller';
+import { ChoicesContext, ChoicesContextValue } from './ChoicesContext';
+
+export const useChoicesContext = <ChoicesType extends RaRecord = RaRecord>(
```
Comment (fzaninotto) [inline packages/ra-core/src/form/choices/useChoicesContext.ts:6]: There is something here that I don't understand: how does passing a `choices` prop override the 3 choices keys in the result? In other terms, how does it NOT break SelectInput and AutocompleteInput when used standalone?

File: packages/ra-core/src/controller/input/useReferenceArrayInputController.ts [hunk 9]
```diff
@@ -239,57 +111,47 @@ export const useReferenceArrayInputController = <
...
         hasNextPage: pageInfo
             ? pageInfo.hasNextPage
             : total != null
-            ? page * perPage < total
+            ? params.page * params.perPage < total
             : undefined,
-        hasPreviousPage: pageInfo ? pageInfo.hasPreviousPage : page > 1,
-        refetch,
-        resource,
-        selectedIds: field.value || EmptyArray,
```
Comment (fzaninotto) [inline packages/ra-core/src/controller/input/useReferenceArrayInputController.ts:292]: I understand the controller no longer manages a "selection", and that it becomes the responsibility of its choild (e.g. DatagridInput) to to it. I think this should be mentioned in the upgrade guide.

File: packages/ra-ui-materialui/src/input/ReferenceArrayInput.spec.js [hunk 10]
```diff
@@ -6,189 +6,48 @@ import {
...
         source: 'tag_ids',
-        translate: x => `*${x}*`,
     };
 
     afterEach(async () => {
         // wait for the getManyAggregate batch to resolve
         await waitFor(() => new Promise(resolve => setTimeout(resolve, 0)));
     });
 
-    it('should display an error if error is defined', () => {
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/input/ReferenceArrayInput.spec.js:36]: we still need a unit test for the error case

File: packages/ra-ui-materialui/src/input/ReferenceArrayInput.tsx [hunk 11]
```diff
@@ -180,127 +130,7 @@ ReferenceArrayInput.defaultProps = {
...
-                  ...getFieldLabelTranslationArgs({
-                      label,
-                      resource,
-                      source,
-                  })
-              )
-            : label;
-
-    if (error) {
-        return <ReferenceError label={translatedLabel} error={error} />;
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/input/ReferenceArrayInput.tsx:249]: what happens in case of error?

File: packages/ra-ui-materialui/src/input/ReferenceInput.spec.tsx [hunk 12]
```diff
@@ -1,87 +1,61 @@
...
-        id: 'post_id',
-        isRequired: false,
-        setFilter: jest.fn(),
-        setPagination: jest.fn(),
-        setSort: jest.fn(),
-        loading: false,
     };
-    const MyComponent = () => <span id="mycomponent" />;
 
-    it('should display an error if error is defined', () => {
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/input/ReferenceInput.spec.tsx:58]: please keep a test for the error case

File: packages/ra-ui-materialui/src/list/BulkActionsToolbar.tsx [hunk 13]
```diff
@@ -48,7 +48,7 @@ export const BulkActionsToolbar = (props: BulkActionsToolbarProps) => {
                         className={BulkActionsToolbarClasses.icon}
                         aria-label={translate('ra.action.unselect')}
                         title={translate('ra.action.unselect')}
-                        onClick={onUnselectItems}
+                        onClick={() => onUnselectItems()}
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/list/BulkActionsToolbar.tsx:51]: why do you need this? This will force a rerender of the button.

File: packages/ra-ui-materialui/src/list/ListGuesser.tsx [hunk 14]
```diff
@@ -79,4 +82,49 @@ ${inferredChild.getRepresentation()}
     return <ListView {...props}>{inferredChild}</ListView>;
 };
 
-ListViewGuesser.propTypes = ListView.propTypes;
+ListViewGuesser.propTypes = {
+    // @ts-ignore-line
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/list/ListGuesser.tsx:86]: why did you need to copy the props? Can't you use Omit instead, which would make the intention clearer?

File: packages/ra-core/src/controller/input/useReferenceParams.ts [hunk 15]
```diff
@@ -0,0 +1,318 @@
...
+    debounce = 500,
+}: ReferenceParamsOptions): [Parameters, Modifiers] => {
+    const [params, setParams] = useState(defaultParams);
+    const tempParams = useRef<ReferenceParams>();
+
+    const requestSignature = [
+        resource,
+        params,
+        filter,
+        JSON.stringify(sort),
```
[the remark this answers is not in the document]
Comment (djhi) [reply packages/ra-core/src/controller/input/useReferenceParams.ts:75]: I copied the [useListParams](https://github.com/marmelab/react-admin/blob/master/packages/ra-core/src/controller/useListParams.ts#L135). Maybe we shoud fix both?

File: packages/ra-ui-materialui/src/list/ListGuesser.tsx (hunk 14 above)
Comment (djhi) [reply packages/ra-ui-materialui/src/list/ListGuesser.tsx:86]: I had a really weird compilation error and haven't found another way to fix it

File: packages/ra-core/src/controller/input/useReferenceParams.ts (hunk 15 above)
[the remark this answers is not in the document]
Comment (djhi) [reply packages/ra-core/src/controller/input/useReferenceParams.ts:75]: Done

File: packages/ra-core/src/form/choices/useChoicesContext.ts (hunk 8 above)
Comment (djhi) [reply packages/ra-core/src/form/choices/useChoicesContext.ts:6]: Because they only use the `allChoices`

File: packages/ra-ui-materialui/src/list/BulkActionsToolbar.tsx (hunk 13 above)
Comment (djhi) [reply packages/ra-ui-materialui/src/list/BulkActionsToolbar.tsx]: Move inside a useCallback

File: packages/ra-ui-materialui/src/list/ListGuesser.tsx (hunk 14 above)
Comment (djhi) [reply packages/ra-ui-materialui/src/list/ListGuesser.tsx]: Fixed

File: packages/ra-ui-materialui/src/input/ReferenceArrayInput.tsx (hunk 11 above)
Comment (djhi) [reply packages/ra-ui-materialui/src/input/ReferenceArrayInput.tsx:249]: We may have to add side effects functions like everywhere else (`onSuccess`, `onError`)?

File: UPGRADE.md [hunk 16]
```diff
@@ -2798,3 +2798,96 @@ import { BooleanInput } from 'react-admin';
 +   onChange={e => { console.log(e.target.checked); }}
 />
 ```
+
+## `<ReferenceInput>` and `<ReferenceArrayInput>` Now Provide A `ChoicesContext`
+
+`<ReferenceInput>` and `<ReferenceArrayInput>` used to inject props to their children. They now provide a `ChoicesContext` with support for sorting, paginating and filtering through methods and properties that match those of the `ListContext`. All their methods now have the same signature. As a result, several changes were made to both components:
+
+### `<ReferenceInput>` No Longer Accepts The `filterToQuery` Prop
```
Comment (WiXSL) [inline UPGRADE.md:2806]: `ReferenceInput` still has `filterToQuery` prop referenced in the code

**Review by WiXSL — CHANGES_REQUESTED** [review]

File: UPGRADE.md (hunk 16 above)
Comment (djhi) [reply UPGRADE.md:2806]: Fixed

Comment (djhi) [reply UPGRADE.md:2806]: btw, adding comments where the issue is would allow github to automatically add the "outdated" flag on the PR comment ;)

File: docs/useChoicesContext.md [hunk 17]
```diff
@@ -0,0 +1,93 @@
...
+    filterValues, // a dictionary of filter values, e.g. { title: 'lorem', nationality: 'fr' }
+    displayedFilters, // a dictionary of the displayed filters, e.g. { title: true, nationality: true }
+    setFilters, // a callback to update the filters, e.g. setFilters(filters, displayedFilters)
+    showFilter, // a callback to show one of the filters, e.g. showFilter('title', defaultValue)
+    hideFilter, // a callback to hide one of the filters, e.g. hideFilter('title')
+    // misc
+    defaultTitle, // the translated title based on the resource, e.g. 'Posts'
+    resource, // the resource name, deduced from the location. e.g. 'posts'
+    refetch, // callback for fetching the list data again
+} = useListContext();
```
Comment (WiXSL) [inline docs/useChoicesContext.md:92]:
```suggestion
} = useChoicesContext();
```

**Review by WiXSL — CHANGES_REQUESTED** [review]

File: docs/useChoicesContext.md [hunk 18]
```diff
@@ -0,0 +1,93 @@
...
+    perPage, // the number of results per page. Defaults to 25
+    setPage, // a callback to change the page, e.g. setPage(3)
+    setPerPage, // a callback to change the number of results per page, e.g. setPerPage(25)
+    hasPreviousPage, // boolean, true if the current page is not the first one
+    hasNextPage, // boolean, true if the current page is not the last one
+    // sorting
+    sort, // a sort object { field, order }, e.g. { field: 'date', order: 'DESC' }
+    setSort, // a callback to change the sort, e.g. setSort({ field: 'name', orfer: 'ASC' })
+    // filtering
+    filterValues, // a dictionary of filter values, e.g. { title: 'lorem', nationality: 'fr' }
```
Comment (WiXSL) [inline docs/useChoicesContext.md:83]: `filter` prop is missing

**Review by WiXSL — CHANGES_REQUESTED** [review]

File: docs/useChoicesContext.md [hunk 19]
```diff
@@ -0,0 +1,93 @@
...
+    displayedFilters, // a dictionary of the displayed filters, e.g. { title: true, nationality: true }
+    setFilters, // a callback to update the filters, e.g. setFilters(filters, displayedFilters)
+    showFilter, // a callback to show one of the filters, e.g. showFilter('title', defaultValue)
+    hideFilter, // a callback to hide one of the filters, e.g. hideFilter('title')
+    // misc
```
Comment (WiXSL) [inline docs/useChoicesContext.md:88]: These misc props are not passed

**Review by WiXSL — CHANGES_REQUESTED** [review]

File: packages/ra-core/src/controller/input/useReferenceArrayInputController.ts [hunk 20]
```diff
@@ -239,57 +111,47 @@ export const useReferenceArrayInputController = <
...
+        setPerPage: paramsModifiers.setPerPage,
+        setSort: paramsModifiers.setSort,
+        showFilter: paramsModifiers.showFilter,
+        source,
+        total: total,
         hasNextPage: pageInfo
             ? pageInfo.hasNextPage
             : total != null
-            ? page * perPage < total
+            ? params.page * params.perPage < total
```
Comment (WiXSL) [inline packages/ra-core/src/controller/input/useReferenceArrayInputController.ts:152]: Well spotted 👍

File: packages/ra-core/src/controller/list/useListController.ts [hunk 21]
```diff
@@ -166,7 +166,7 @@ export const useListController = <RecordType extends RaRecord = any>(
         hasNextPage: pageInfo
             ? pageInfo.hasNextPage
             : total != null
-            ? query.page * perPage < total
+            ? query.page * query.perPage < total
```
Comment (WiXSL) [inline packages/ra-core/src/controller/list/useListController.ts:169]: Well spotted 👍

**Review by WiXSL — CHANGES_REQUESTED** [review]
**ReferenceArrayInput.tsx**
JsDocs need to be upgraded to refer to `ChoicesContext` rather than `choices` prop

https://github.com/marmelab/react-admin/blob/07faec5d247df90d11b27b1abe055b7f4c72ea68/packages/ra-ui-materialui/src/input/ReferenceArrayInput.tsx#L26

**Review by WiXSL — CHANGES_REQUESTED** [review]
**useReferenceArrayInputController.ts**
JsDocs example needs upgrade

https://github.com/marmelab/react-admin/blob/07faec5d247df90d11b27b1abe055b7f4c72ea68/packages/ra-core/src/controller/input/useReferenceArrayInputController.ts#L14

**Review by WiXSL — CHANGES_REQUESTED** [review]
**useGetManyAggregate.ts**
JsDocs needs upgrade

https://github.com/marmelab/react-admin/blob/07faec5d247df90d11b27b1abe055b7f4c72ea68/packages/ra-core/src/dataProvider/useGetManyAggregate.ts#L55

**Comment by djhi** [comment]
> useGetManyAggregate.ts
JsDocs needs upgrade

Why ?

**Comment by WiXSL** [comment]
> > useGetManyAggregate.ts
> > JsDocs needs upgrade
> 
> Why ?

At least loading isn't there, it should be isFetching:

```
const { data, isFetching, error } = useGetManyAggregate('tags', { ids: record.tagIds });

File: docs/useChoicesContext.md [hunk 22]
```diff
@@ -0,0 +1,95 @@
...
+
+```jsx
+// in src/comments/PostInput.js
+import { AutocompleteInput, useChoicesContext } from 'react-admin';
+
+export const PostInput = (props) => {
+    const { setFilters } = useChoicesContext();
+
+    const handleCheckboxChange = (event, checked) => {
+        setFilters({ published: checked });
```
Comment (WiXSL) [inline docs/useChoicesContext.md:28]:
```suggestion
        setFilters({ published: checked }, displayedFilters);
```

File: docs/useChoicesContext.md [hunk 23]
```diff
@@ -0,0 +1,95 @@
...
+## Usage
+
+Call `useChoicesContext` in a component, then use this component as a descendant of a `ReferenceInput` or `ReferenceArrayInput` component.
+
+```jsx
+// in src/comments/PostInput.js
+import { AutocompleteInput, useChoicesContext } from 'react-admin';
+
+export const PostInput = (props) => {
+    const { setFilters } = useChoicesContext();
```
Comment (WiXSL) [inline docs/useChoicesContext.md:25]:
```suggestion
    const { setFilters, displayedFilters } = useChoicesContext();
```

**Review by WiXSL — CHANGES_REQUESTED** [review]

**Review by WiXSL — APPROVED** [review]

File: UPGRADE.md (hunk 16 above)
Comment (WiXSL) [inline UPGRADE.md:2806]: The problem now is that there is no reference in the docs to the `filterToQuery` prop. It should be in `AutocompleteInput` and `SelectInput` docs

**Review by WiXSL — CHANGES_REQUESTED** [review]

**Comment by fzaninotto** [comment]
@WiXSL Sorry I saw your comment after merging. Would you like to make the change in another PR?
